README.md
A nodejs implementation of zenmoney api:
https://github.com/zenmoney/ZenPlugins/wiki/ZenMoney-API

## Installation

```bash
npm install zenmoney-api
```

or

```bash
yarn add zenmoney-api
```

## Usage example

```js
import { zenmoneyApi } from 'zenmoney-api';

const username = '...';
const password = '...';
const apiKey = '...';
const apiSecret = '...';

(async () => {
  try {
    // you can also use `zenmoneyApi.setToken('token')`
    // if you already have one
    await zenmoneyApi.authorize({
      username,
      password,
      apiKey,
      apiSecret,
    });

    const diff = await zenmoneyApi.diff({
      // fetch prev two months
      serverTimestamp: Math.round(
        (Date.now() - 60 * 24 * 60 * 60 * 1000) / 1000,
      ),
    });

    // strip arrays to only 3 items
    Object.values(diff).forEach((item) => {
      if (Array.isArray(item)) {
        item.splice(0, item.length - 3);
      }
    });

    console.dir(diff, { depth: null });

    const suggestion = await zenmoneyApi.suggest({
      payee: 'McDonalds',
    });

    console.log('Suggestion for McDonalds');
    console.dir(suggestion, { depth: null });
  } catch (error) {
    console.log('Error requesting diff', {
      error,
      request: error.request,
      response: error.response,
    });

    process.exit(1);
  }

  process.exit(0);
})();
```

## TODO

- support oauth authentication
- automatic token invalidation and update
- improve api error checking


CHANGELOG.md
# Changelog

## [2.0.0](https://github.com/SleepWalker/zenmoney-api/compare/v1.0.0...v2.0.0) (2023-08-28)

### ⚠ BREAKING CHANGES

- bump deps to latest versions and fix vulnerabilities. BREAKING: Node >= 16 is
  required

### Features

- expose getToken/setToken methods (resolves
  [#18](https://github.com/SleepWalker/zenmoney-api/issues/18))
  ([340650f](https://github.com/SleepWalker/zenmoney-api/commit/340650feecda267f4d49fd93faa960e37a1e220b))

### Code Refactoring

- bump deps to latest versions and fix vulnerabilities. BREAKING: Node >= 16 is
  required
  ([5efc89f](https://github.com/SleepWalker/zenmoney-api/commit/5efc89fb96a0b8020020254ac767ae070d595c63))
- migrate to npm
  ([5ffdbe9](https://github.com/SleepWalker/zenmoney-api/commit/5ffdbe9564c21f015e2d5b78f20e4ff5896e885d))
- migrate to semantic-release
  ([992dbe5](https://github.com/SleepWalker/zenmoney-api/commit/992dbe5bc68506de543d2aed81ae488b64a4227e))


src/index.test.ts
import { zenmoneyApi } from './';

it('should export api object', () => {
  expect(zenmoneyApi).toEqual(expect.any(Object));
});


src/auth.ts
import assert from 'assert';

import type { AuthData } from './types';
import client from './httpClient';

interface TokenResponse {
  access_token: string;
  token_type: 'bearer';
  expires_in: number; // seconds,
  refresh_token: string;
}

class AuthApi {
  async getToken({
    username,
    password,
    apiKey,
    apiSecret,
  }: {
    username: string;
    password: string;
    apiKey: string;
    apiSecret: string;
  }) {
    assert(username, 'username is required for request');
    assert(password, 'password is required for request');
    assert(apiKey, 'apiKey is required for request');
    assert(apiSecret, 'apiSecret is required for request');

    const redirectUrl = 'http://0.0.0.0';
    await this.collectAuthCookies({
      redirectUrl,
      apiKey,
    });

    const code = await this.getAuthorizeCode({
      username,
      password,
      redirectUrl,
      apiKey,
    });

    const token = await this.requestToken({
      code,
      apiKey,
      apiSecret,
      redirectUrl,
    });

    return token;
  }

  private async collectAuthCookies({
    redirectUrl,
    apiKey,
  }: {
    redirectUrl: string;
    apiKey: string;
  }): Promise<void> {
    await client.get('oauth2/authorize', {
      responseType: 'text',
      searchParams: {
        response_type: 'code',
        client_id: apiKey,
        redirect_uri: redirectUrl,
      },
    });
  }

  private async getAuthorizeCode({
    username,
    password,
    redirectUrl,
    apiKey,
  }: {
    username: string;
    password: string;
    redirectUrl: string;
    apiKey: string;
  }): Promise<string> {
    const {
      headers: { location: urlWithCode },
    } = await client.post('oauth2/authorize', {
      responseType: 'text',
      searchParams: {
        response_type: 'code',
        client_id: apiKey,
        redirect_uri: redirectUrl,
      },
      form: {
        username: username,
        password: password,
        auth_type_password: 'Sign in',
      },
      followRedirect: false,
    });

    const query = new URLSearchParams(String(urlWithCode).split('?')[1]);

    const code = query.get('code');

    if (!code) {
      throw new Error('Can not get auth code');
    }

    return code;
  }

  private async requestToken({
    code,
    redirectUrl,
    apiKey,
    apiSecret,
  }: {
    code: string;
    redirectUrl: string;
    apiKey: string;
    apiSecret: string;
  }): Promise<AuthData> {
    const {
      body: {
        access_token: accessToken,
        token_type: tokenType,
        refresh_token: refreshToken,
        expires_in: expiresIn,
      },
    } = await client.post<TokenResponse>('oauth2/token', {
      form: {
        grant_type: 'authorization_code',
        client_id: apiKey,
        client_secret: apiSecret,
        code: code,
        redirect_uri: redirectUrl,
      },
    });

    return {
      tokenType,
      accessToken,
      refreshToken,
      expiresIn,
    };
  }
}

export default new AuthApi();


src/httpClient.ts
import got from 'got';
import { CookieJar } from 'tough-cookie';

const defaultContext: {
  token: null | string;
} = {
  token: null,
};

function proxy(org, proxyFn) {
  return new Proxy(proxyFn(org), {
    get: (obj, prop) => (prop in obj ? obj[prop] : org[prop]),
  });
}

const cookieJar = proxy(new CookieJar(), (obj: CookieJar) => {
  return {
    setCookie: async (rawCookie: string, url: string) =>
      obj.setCookie(rawCookie, url),
    getCookieString: async (url: string) => obj.getCookieString(url),
  };
});

const client = got.extend({
  prefixUrl: 'https://api.zenmoney.ru',
  responseType: 'json',
  retry: {
    limit: 4,
    methods: [
      'GET',
      // currently all POST requests are idempotent and can be safely retried
      'POST',
    ],
  },
  cookieJar,
  hooks: {
    beforeRequest: [
      (options) => {
        const { token } = defaultContext;

        if (token) {
          options.headers.authorization = `Bearer ${token}`;
        }
      },
    ],
  },
});

export const setToken = (token: string): void => {
  defaultContext.token = token;
};

export default client;


src/zenmoneyApi.ts
import type { DiffObject, Transaction, UnixTimestamp } from './types';
import auth from './auth';
import client, { setToken } from './httpClient';

const unixNow = (): UnixTimestamp => Math.round(Date.now() / 1000);

export interface ApiAuthData {
  username: string;
  password: string;
  apiKey: string;
  apiSecret: string;
}

class ZenmoneyApi {
  async getToken(authData: ApiAuthData) {
    return auth.getToken(authData);
  }

  setToken(token: string) {
    setToken(token);
  }

  async authorize(authData: ApiAuthData) {
    const { accessToken: token } = await this.getToken(authData);

    this.setToken(token);
  }

  async diff({
    currentClientTimestamp = unixNow(),
    serverTimestamp = 0,
  }: DiffObject = {}): Promise<DiffObject> {
    const resp = await client.post<DiffObject>('v8/diff', {
      json: {
        currentClientTimestamp,
        serverTimestamp,
      },
    });

    return resp.body;
  }

  async suggest<T extends Partial<Transaction>>(transaction: T): Promise<T> {
    const resp = await client.post<T>('v8/suggest', {
      json: transaction,
    });

    return resp.body;
  }
}

export const zenmoneyApi = new ZenmoneyApi();


src/types.ts
/**
 * @see https://github.com/zenmoney/ZenPlugins/wiki/ZenMoney-API
 */

export interface AuthData {
  accessToken: string;
  tokenType: 'bearer';
  expiresIn: number; // seconds,
  refreshToken: string;
}

export type UnixTimestamp = number;

/**
 * System entities (read only)
 */

/**
 * Currency data and rates
 */
export interface Instrument {
  id: number;
  title: string;
  shortTitle: string; // 'USD'
  symbol: string; // '$'
  rate: number; // float
  changed: UnixTimestamp;
}

/**
 * Bank or any other payment company that may control some financial accounts
 */
export interface Company {
  id: number;
  title: string;
  fullTitle: string;
  www: string;
  country: string;
  changed: UnixTimestamp;
}

export interface Country {
  id: number;
  title: string;
  domain: string; // 'ru' | 'ua' | 'us'
  currency: Instrument['id'];
}

export interface User {
  id: number;
  login: string | null;

  /**
   * Main user's currency. It will be used for calculations
   * and reports inside system
   */
  currency: Instrument['id'];

  /**
   * Parent (admin) user id for the family family accounting.
   * This user can control which user can be in family.
   * Parent user has parent === null
   */
  parent: User['id'] | null;
  changed: UnixTimestamp;
}

/**
 * User entities (can be edited through API)
 */

export interface Account {
  id: string; // UUID
  changed: UnixTimestamp;
  user: User['id'];
  role: User['id'] | null;
  instrument: Instrument['id'] | null;
  company: Company['id'] | null;
  type: 'cash' | 'ccard' | 'checking' | 'loan' | 'deposit' | 'emoney' | 'debt';
  title: string;
  /**
   * An array of the last 4 digits of the corresponding bank cards or account numbers
   */
  syncID: string[] | null;

  /**
   * Current account balance
   */
  balance: number | null;

  /**
   * Account balance (or loan body) on opening date
   */
  startBalance: number | null;

  /**
   * Credit limit for ccard
   *
   * @return  {[type]}  [return description]
   */
  creditLimit: number | null; // >= 0

  inBalance: boolean;
  savings: boolean | null;

  /**
   * Whether this account's balance should be corrected based on SMS data
   */
  enableCorrection: boolean;
  enableSMS: boolean;
  archive: boolean;

  // All next fields may be null in case
  // for all account types except 'loan' and 'deposit'
  capitalization: boolean;
  percent: number; // >= 0 && < 100
  startDate: string; // 'yyyy-MM-dd'
  endDateOffset: number;
  endDateOffsetInterval: 'day' | 'week' | 'month' | 'year';
  payoffStep: number | null;
  payoffInterval: 'month' | 'year' | null;
}

/**
 * Category is named Tag in an official documentation
 */
export interface Category {
  id: string; // UUID
  user: User['id'];

  title: string;
  /**
   * Parent category. It is allowed only 1 level nesting
   */
  parent: Category['id'] | null;
  /**
   * Icon id
   */
  icon: string | null;
  /**
   * Icon color
   *
   * @return  {[type]}  [return description]
   */
  color: number | null;
  /**
   * Icon picture url
   *
   * Calculation:
   * alpha, red, green, blue 0 <= 255.
   * unsigned long color = (a << 24) + (r << 16) + (g << 8) + (b << 0)
   */
  picture: string | null;

  showIncome: boolean;
  showOutcome: boolean;
  /**
   * Whether category is taking part in income budget
   */
  budgetIncome: boolean;
  /**
   * Whether category is taking part in outcome budget
   */
  budgetOutcome: boolean;
  /**
   * Whether expenses in this category are required.
   * `true` or `null` means required
   */
  required: boolean | null;

  changed: UnixTimestamp;
}

/**
 * The Merchant of operation
 *
 * Unlike operation payee merchant is stored and than can be suggested during
 * transaction creation. Is also listed on payee and payer page
 */
export interface Merchant {
  id: string; // UUID
  user: User['id'];
  title: string;
  changed: UnixTimestamp;
}

interface TransactionSpecification {
  incomeInstrument: Instrument['id'];
  incomeAccount: Account['id'];
  income: number; // >= 0
  outcomeInstrument: Instrument['id'];
  outcomeAccount: Account['id'];
  outcome: number; // >= 0
}

interface ReminderBase extends TransactionSpecification {
  id: string; // UUID
  user: User['id'];

  tag: Category['id'][] | null;
  merchant: Merchant['id'] | null;
  payee: string | null;
  comment: string | null;

  notify: boolean;

  changed: UnixTimestamp;
}

/**
 * Reminder operation settings
 *
 * @see https://github.com/zenmoney/ZenPlugins/wiki/ZenMoney-API#reminder
 */
export interface Reminder extends ReminderBase {
  /**
   * Repeatability interval. `null` if transaction does not repeat
   */
  interval: 'day' | 'week' | 'month' | 'year' | null;
  /**
   * The step size of interval. E.g. 2 days
   */
  step: number | null; // >= 0
  /**
   * The points inside step, where reminders are created
   */
  points: number[] | null; // >= 0 && < step
  /**
   * Reminder start date
   */
  startDate: string; // 'yyyy-MM-dd'
  /**
   * Reminder end date (inclusive).
   * If `null` — repeat forever
   */
  endDate: string | null; // 'yyyy-MM-dd'
}

/**
 * Actual reminder operation that was created based on Reminder configuration
 */
export interface ReminderMarker extends ReminderBase {
  reminder: Reminder['id'];

  /**
   * planned - planned operation
   * processed - completed operation (transaction was created)
   * deleted - the reminder marker was deleted
   */
  state: 'planned' | 'processed' | 'deleted';
  date: string; // 'yyyy-MM-dd'
}

export interface Transaction extends TransactionSpecification {
  id: string; // UUID
  user: User['id'];
  deleted: boolean;
  hold: boolean | null;

  tag: Category['id'][] | null;
  merchant: Merchant['id'] | null;
  payee: string | null;
  originalPayee: string | null;
  comment: string | null;

  date: string; // 'yyyy-MM-dd'

  mcc: number | null;

  reminderMarker: ReminderMarker['id'] | null;

  opIncome: number | null; // >= 0
  opIncomeInstrument: Instrument['id'] | null;
  opOutcome: number | null; // >= 0
  opOutcomeInstrument: Instrument['id'] | null;

  latitude: number | null; // >= -90  && <= 90
  longitude: number | null; // >= -180 && <= 180

  changed: UnixTimestamp;
  created: UnixTimestamp;
}

/**
 * Budget for the month
 *
 * NOTE: To remove a budget for specified month you need to set lock to false
 * and set outcome/income to zero
 */
export interface Budget {
  user: User['id'];

  /**
   * Budget category.
   * If `null` — uncategorized operations
   * If '00000000-0000-0000-0000-000000000000' — total budget for month
   */
  tag: Category['id'] | null | '00000000-0000-0000-0000-000000000000';
  date: string; // 'yyyy-MM-dd'

  income: number;
  /**
   * If `true` — means that income summ also includes all planned operations
   */
  incomeLock: boolean;
  outcome: number;
  /**
   * If `true` — means that outcome summ also includes all planned operations
   */
  outcomeLock: boolean;

  changed: UnixTimestamp;
}

export type EntityName =
  | 'instrument'
  | 'country'
  | 'company'
  | 'user'
  | 'account'
  | 'tag'
  | 'budget'
  | 'merchant'
  | 'reminder'
  | 'reminderMarker'
  | 'transaction';

export interface DiffObject {
  /**
   * The time on a client for time correction
   */
  currentClientTimestamp?: UnixTimestamp;

  /**
   * The time of the last synchronization
   * To start sync from scratch — set `0`
   *
   * You should save this value from API response to use during the next sync
   */
  serverTimestamp?: UnixTimestamp;

  /**
   * Which entities should the server return as of it was the initial sync
   */
  forceFetch?: EntityName[];

  instrument?: Instrument[];
  country?: Country[];
  company?: Company[];
  user?: User[];
  account?: Account[];
  tag?: Category[];
  budget?: Budget[];
  merchant?: Merchant[];
  reminder?: Reminder[];
  reminderMarker?: ReminderMarker[];
  transaction?: Transaction[];

  deletion?: {
    id: string; // entity id
    object: EntityName;
    stamp: number;
    user: number;
  }[];
}
